dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/relation/RelationModelManager.java

```diff
@@ -0,0 +1,596 @@
+                                    try {
+                                        return Integer.parseInt(s.substring(s.lastIndexOf("#") + 1));
+                                    } catch (NumberFormatException e) {
+                                        return Integer.MIN_VALUE;
+                                    }
+                                  })
+                                  .filter(i -> i != Integer.MIN_VALUE)
+                                  .max(Comparator.naturalOrder()).orElse(0);
+            if (defIndex != 0) {
+                redefineds.put(defIndex, r);
+            }
+        }
+        return redefineds.entrySet().stream().max(Map.Entry.comparingByKey())
+                         .map(Map.Entry::getValue).orElse(wmm.getRelation(name));
+    }
+
+    private void extractRelations(List<String> relationNames) {
+        Set<Relation> relsToExtract = new HashSet<>();
+        for (String name : relationNames) {
+            Relation r = getRelationWithName(name);
+            if (r == null) {
+                logger.warn("Relation with the name {} does not exist", name);
+                continue;
+            }
+            relsToExtract.add(r);
+            createModel(r, name);
+        }
+
+        Set<RelationGraph> relGraphs = new HashSet<>();
+        DependencyGraph<Relation> dependencyGraph = DependencyGraph.from(relsToExtract);
+
+        for (Set<DependencyGraph<Relation>.Node> component : dependencyGraph.getSCCs()) {
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = new RecursiveGraph();
+                    recGraph.setName(r.getNameOrTerm() + "_rec");
+                    relGraphs.add(recGraph);
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
+
+        Set<CAATPredicate> predicates = new HashSet<>(relGraphs);
+        PredicateHierarchy hierarchy = new PredicateHierarchy(predicates);
+        hierarchy.initializeToDomain(domain);
+
+        // Populate graphs of base relations.
+        for (CAATPredicate basePred : hierarchy.getBasePredicates()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) basePred);
+            try {
+                r.getDefinition().accept(graphPopulator);
+            } catch (UnsupportedOperationException e) {
+                graphPopulator.populateDynamicDefaultGraph(r);
+            }
+        }
```

Comment:
Is this to have some default fallback when some visitor is not implemented? If so, this needs to be at least documented